app/controllers/resources_controller.rb

```diff
@@ -127,26 +127,43 @@ def file_identifier(object_id, resource_id)
     "#{ID_NAMESPACE}/file/#{object_id.delete_prefix('druid:')}-#{resource_id}/#{SecureRandom.uuid}"
   end
 
-  # rubocop:disable Metrics/AbcSize
-  def decorate_file(file:, version:, external_id: nil)
-    blob = blob_for_signed_id(file.delete(:externalIdentifier), file[:filename])
-    file[:externalIdentifier] = external_id if external_id
-    file[:version] = version
+  def metadata_for_blob(blob, file)
+    file.delete(:externalIdentifier)
     file[:size] = blob.byte_size
     file[:hasMimeType] = blob.content_type || 'application/octet-stream'
     declared_md5 = file[:hasMessageDigests].find { |digest| digest.fetch(:type) == 'md5' }.fetch(:digest)
     calculated_md5 = base64_to_hexdigest(blob.checksum)
     raise BlobError, "MD5 mismatch for #{file[:filename]}" if declared_md5 != calculated_md5
   end
-  # rubocop:enable Metrics/AbcSize
+
+  def metadata_for_file(globus_file, file)
+    return unless File.exist?(globus_file)
+
+    file[:size] = File.size(globus_file)
+  end
+
+  def decorate_file(file:, version:, external_id: nil)
+    if signed_id?(file[:externalIdentifier])
+      blob = blob_for_signed_id(file.delete(:externalIdentifier), file[:filename])
+      metadata_for_blob(blob, file)
+    else
+      globus_file = file_from_globus(file.delete(:externalIdentifier), file[:filename])
+      metadata_for_file(globus_file, file)
+    end
+
+    # Set file params post-processing
+    file[:externalIdentifier] = external_id if external_id
+    file[:version] = version
+  end
 
   # Decorates the provided FileSets with the information we have in the ActiveStorage table.
   # externalIdentifier is also removed from the request.
   def decorate_request_file_sets(model_params)
     file_sets(model_params).each do |fileset|
       fileset[:version] = 1
       fileset.dig(:structural, :contains).each do |file|
-        decorate_file(file: file, version: 1)
+        external_identifier = signed_id?(file[:externalIdentifier]) ? nil : file[:externalIdentifier]
```

Comment:
I see we check if the external identifer is a signed_id here, and then we check it again immediately after (in the first line of `#decorate_file`). If we're checking again so soon and we still have the file object (and thus access to the external ID), do you feel we need both calls?

Reply:
Good catch, probably cruft/oversight. I'll take a look. Thanks :-)